components/region_cache_memory_engine/src/memory_controller.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2024 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+
+use crossbeam::utils::CachePadded;
+
+use crate::{engine::SkiplistEngine, write_batch::NODE_OVERHEAD_SIZE_EXPECTATION};
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum MemoryUsage {
+    NormalUsage(usize),
+    SoftLimitReached(usize),
+    // usize here means the current memory usage and it's the usize in it adding with the memory
+    // acquiring exceeds the hard limit
+    HardLimitReached(usize),
+}
+
+#[derive(Debug)]
+struct HotData {
+    // Allocated memory for keys and values (node overhead is not included)
+    allocated: AtomicUsize,
+    // The number of writes that are buffered but not yet written.
+    pending_node_count: AtomicUsize,
```

Comment:
I think generally speaking `pending_node_count` is not needed.
We can use self.skiplist_engine.node_count() directly, as the difference is neglectable.
This can save some CPU usage and also make the code a bit simpler---so we don't need HotData struct, just AtomicUSize `allocated`